File: crates/buquet/examples/worker.py
#!/usr/bin/env python3
"""
Example worker that processes orders.

Run with:
    cd crates/buquet
    uv run python examples/worker.py
"""
# ruff: noqa: T201 S311

from __future__ import annotations

import asyncio
import random

from buquet import RetryableError, Worker, WorkerRunOptions, connect

# Simulated product catalog
PRODUCTS: dict[str, dict[str, object]] = {
    "widget": {"name": "Widget", "price": 19.99},
    "gadget": {"name": "Gadget", "price": 5.99},
    "gizmo": {"name": "Gizmo", "price": 14.99},
    "doohickey": {"name": "Doohickey", "price": 29.99},
    "thingamajig": {"name": "Thingamajig", "price": 9.99},
}

# Constants for simulation
TRANSIENT_FAILURE_RATE = 0.1


async def main() -> None:
    print("[worker] Starting order processing worker...")

    # Connect to the queue
    queue = await connect()

    # Create a worker that handles all 16 shards
    shards = [f"{i:x}" for i in range(16)]  # ["0", "1", ..., "f"]
    worker = Worker(queue, "order-worker-1", shards)

    @worker.task("process_order")
    async def process_order(data: dict[str, object]) -> dict[str, object]:
        """Process an order by calculating totals and simulating work."""
        order_id = str(data["order_id"])
        items = list(data["items"])  # type: ignore[arg-type]

        print(f"[worker] Processing order {order_id}...")

        # Calculate line items
        line_items = []
        subtotal = 0.0

        for item in items:
            product_id = item["product_id"]
            quantity = item["quantity"]

            if product_id not in PRODUCTS:
                msg = f"Unknown product: {product_id}"
                raise ValueError(msg)

            product = PRODUCTS[product_id]
            price = float(product["price"])  # type: ignore[arg-type]
            name = str(product["name"])
            line_total = price * quantity
            subtotal += line_total

            line_items.append(
                {
                    "product_id": product_id,
                    "name": name,
                    "quantity": quantity,
                    "unit_price": price,
                    "total": round(line_total, 2),
                }
            )

            print(f"[worker]   - {quantity}x {name} @ ${price:.2f} = ${line_total:.2f}")

        # Simulate some processing time (0.2-0.8 seconds)
        process_time = random.uniform(0.2, 0.8)
        await asyncio.sleep(process_time)

        # Simulate occasional transient failures (10% chance)
        if random.random() < TRANSIENT_FAILURE_RATE:
            print(f"[worker] Transient error on {order_id}, will retry...")
            msg = "Payment gateway temporarily unavailable"
            raise RetryableError(msg)

        subtotal = round(subtotal, 2)
        print(f"[worker]   - Subtotal: ${subtotal:.2f}")
        print(f"[worker] Completed {order_id} in {process_time:.1f}s")

        return {
            "order_id": order_id,
            "line_items": line_items,
            "subtotal": subtotal,
            "status": "processed",
        }

    print(f"[worker] Registered handlers: {worker.registered_task_types()}")
    print("[worker] Polling for tasks...")

    # Run the worker (polls every 500ms)
    await worker.run(WorkerRunOptions(poll_interval_ms=500))


if __name__ == "__main__":
    asyncio.run(main())

File: examples/worker.rs
//! Example Rust worker that processes orders.
//!
//! Run with:
//!     cargo run -p buquet --example worker
//!
//! This example demonstrates:
//! - Connecting to buquet via S3 (auto-loads .buquet.toml config)
//! - Implementing `TaskHandler` trait
//! - Registering handlers with a Worker
//! - Running the worker loop with graceful shutdown
//! - Polyglot interop (can consume tasks from Python producer)

use async_trait::async_trait;
use buquet::models::TaskError;
use buquet::queue::connect;
use buquet::worker::{
    shutdown_signal, wait_for_shutdown_signal, IndexMode, RunnerConfig, TaskHandler, Worker,
};
use serde_json::{json, Value};
use std::time::Duration;

// Simulated product catalog
fn get_product(product_id: &str) -> Option<(&'static str, f64)> {
    match product_id {
        "widget" => Some(("Widget", 19.99)),
        "gadget" => Some(("Gadget", 5.99)),
        "gizmo" => Some(("Gizmo", 14.99)),
        "doohickey" => Some(("Doohickey", 29.99)),
        "thingamajig" => Some(("Thingamajig", 9.99)),
        _ => None,
    }
}

/// Handler for processing order tasks
struct ProcessOrderHandler;

#[async_trait]
impl TaskHandler for ProcessOrderHandler {
    fn task_type(&self) -> &'static str {
        "process_order"
    }

    async fn handle(&self, input: Value) -> Result<Value, TaskError> {
        let order_id = input["order_id"].as_str().unwrap_or("unknown").to_string();

        let items = input["items"]
            .as_array()
            .ok_or_else(|| TaskError::Permanent("Missing items array".to_string()))?;

        println!("[worker] Processing order {order_id}...");

        let mut line_items = Vec::new();
        let mut subtotal = 0.0;

        for item in items {
            let product_id = item["product_id"]
                .as_str()
                .ok_or_else(|| TaskError::Permanent("Missing product_id".to_string()))?;

            let quantity = item["quantity"]
                .as_u64()
                .ok_or_else(|| TaskError::Permanent("Missing quantity".to_string()))?
                as f64;

            let (name, price) = get_product(product_id)
                .ok_or_else(|| TaskError::Permanent(format!("Unknown product: {product_id}")))?;

            let line_total = price * quantity;
            subtotal += line_total;

            println!(
                "[worker]   - {}x {} @ ${:.2} = ${:.2}",
                quantity as u64, name, price, line_total
            );

            line_items.push(json!({
                "product_id": product_id,
                "name": name,
                "quantity": quantity as u64,
                "unit_price": price,
                "total": (line_total * 100.0).round() / 100.0,
            }));
        }

        // Simulate some processing time
        tokio::time::sleep(Duration::from_millis(200)).await;

        // Simulate occasional transient failures (10% chance)
        if rand::random::<f64>() < 0.1 {
            println!("[worker] Transient error on {order_id}, will retry...");
            return Err(TaskError::Retryable(
                "Payment gateway temporarily unavailable".to_string(),
            ));
        }

        let subtotal = (subtotal * 100.0).round() / 100.0;
        println!("[worker]   Subtotal: ${subtotal:.2}");
        println!("[worker] Completed {order_id}");

        Ok(json!({
            "order_id": order_id,
            "line_items": line_items,
            "subtotal": subtotal,
            "status": "processed",
        }))
    }
}

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    // Initialize tracing
    tracing_subscriber::fmt::init();

    println!("[worker] Rust worker starting...");

    // Connect using automatic config resolution (env vars / .buquet.toml)
    let queue = connect(None).await?;
    println!("[worker] Connected to bucket: {}", queue.client().bucket());

    // Create worker with all 16 shards
    let shards: Vec<String> = (0..16).map(|i| format!("{i:x}")).collect();
    let worker_id = format!("rust-worker-{}", &uuid::Uuid::new_v4().to_string()[..8]);

    let mut worker = Worker::new(queue, &worker_id, shards);

    // Register handler
    worker.register_handler(Box::new(ProcessOrderHandler));

    println!("[worker] Worker ID: {worker_id}");
    println!("[worker] Registered handlers: {:?}", worker.handlers());
    println!("[worker] Polling for tasks (Ctrl+C to stop)...");
    println!();

    // Setup shutdown signal
    let (shutdown_tx, shutdown_rx) = shutdown_signal();
    tokio::spawn(wait_for_shutdown_signal(shutdown_tx));

    // Run the worker
    let runner_config = RunnerConfig {
        polling: buquet::worker::PollingStrategy::fixed(500),
        index_mode: IndexMode::Hybrid,
        ..Default::default()
    };

    worker.run(runner_config, shutdown_rx).await?;

    println!("[worker] Shutdown complete");

    Ok(())
}
